examples/core/rgcn/hetero_rgcn.py

```diff
@@ -0,0 +1,654 @@
+        self.out_size = out_size
+        self.ntypes = ntypes
+        self.relation_names = relation_names
+        self.activation = activation
+
+        ########################################################################
+        # (HIGHLIGHT) HeteroGraphConv is a graph convolution operator over
+        # heterogeneous graphs. A dictionary is passed where the key is the
+        # relation name and the value is the instance of GraphConv. norm="right"
+        # is to divide the aggregated messages by each node’s in-degrees, which
+        # is equivalent to averaging the received messages. weight=False and
+        # bias=False as we will use our own weight matrices defined later.
+        ########################################################################
+        self.conv = dglnn.HeteroGraphConv(
+            {
+                rel: dglnn.GraphConv(
+                    in_size, out_size, norm="right", weight=False, bias=False
+                )
+                for rel in relation_names
+            }
+        )
+
+        # Create a separate Linear layer for each relationship. Each
+        # relationship has its own weights which will be applied to the node
+        # features before performing convolution.
+        self.weight = nn.ModuleDict(
+            {
+                rel_name: nn.Linear(in_size, out_size, bias=False)
+                for rel_name in self.relation_names
+            }
+        )
+
+        # Create a separate Linear layer for each node type for self-loop
+        # connections, i.e., edges that connect nodes to themselves. These
+        # weights help the model to carry information from a node to itself
+        # across layers.
+        self.loop_weights = nn.ModuleDict(
+            {
+                ntype: nn.Linear(in_size, out_size, bias=True)
+                for ntype in self.ntypes
+            }
+        )
+
+        self.dropout = nn.Dropout(dropout)
+        # Initialize parameters of the model.
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        for layer in self.weight.values():
+            layer.reset_parameters()
+
+        for layer in self.loop_weights.values():
+            layer.reset_parameters()
+
+    def forward(self, g, inputs):
+        """
+        Parameters
+        ----------
+        g : DGLGraph
+            Input graph.
+        inputs : dict[str, torch.Tensor]
+            Node feature for each node type.
+
+        Returns
+        -------
+        dict[str, torch.Tensor]
+            New node features for each node type.
+        """
+        # Create a deep copy of the graph g with features saved in local
+        # frames to prevent side effects from modifying the graph.
+        g = g.local_var()
+
+        # Create a dictionary of weights for each relationship. The weights
+        # are retrieved from the Linear layers defined earlier.
+        weight_dict = {
+            rel_name: {"weight": self.weight[rel_name].weight.T}
```

Comment:
Why the `.T` is necessary here? Can we change the matrix shape when creating the linear modules?

Reply:
Did you mean:
```python
self.weight = nn.ModuleDict(
    {
        # swap in_size and out_size.
        rel_name: nn.Linear(out_size, in_size, bias=False)
        for rel_name in self.relation_names
    }
)
```
So that we didn't need to transpose.